packages/graphiql-react/src/editor/completion.ts

```diff
@@ -18,28 +18,122 @@ export function onHasCompletion(
+      (ctx: IHint, el: HTMLDivElement) => {
         // Only the first time (usually when the hint UI is first displayed)
         // do we create the information nodes.
         if (!information) {
-          const hintsUl = el.parentNode as Node & ParentNode;
+          const hintsUl = el.parentNode as HTMLUListElement & ParentNode;
 
           // This "information" node will contain the additional info about the
           // highlighted typeahead option.
           information = document.createElement('div');
           information.className = 'CodeMirror-hint-information';
-          information.addEventListener('click', onClickHintInformation);
           hintsUl.appendChild(information);
 
-          // This "deprecation" node will contain info about deprecated usage.
+          const header = document.createElement('header');
+          header.className = 'CodeMirror-hint-information-header';
+          information.appendChild(header);
+
+          fieldName = document.createElement('span');
+          fieldName.className = 'CodeMirror-hint-information-field-name';
+          header.appendChild(fieldName);
+
+          typeNamePill = document.createElement('span');
+          typeNamePill.className = 'CodeMirror-hint-information-type-name-pill';
+          header.appendChild(typeNamePill);
+
+          typeNamePrefix = document.createElement('span');
+          typeNamePill.appendChild(typeNamePrefix);
+
+          typeName = document.createElement('a');
+          typeName.className = 'CodeMirror-hint-information-type-name';
+          typeName.href = 'javascript:void 0'; // eslint-disable-line no-script-url
+          typeName.addEventListener('click', onClickHintInformation);
+          typeNamePill.appendChild(typeName);
+
+          typeNameSuffix = document.createElement('span');
+          typeNamePill.appendChild(typeNameSuffix);
+
+          description = document.createElement('div');
+          description.className = 'CodeMirror-hint-information-description';
+          information.appendChild(description);
+
           deprecation = document.createElement('div');
-          deprecation.className = 'CodeMirror-hint-deprecation';
-          hintsUl.appendChild(deprecation);
+          deprecation.className = 'CodeMirror-hint-information-deprecation';
+          information.appendChild(deprecation);
+
+          const deprecationLabel = document.createElement('span');
+          deprecationLabel.className =
+            'CodeMirror-hint-information-deprecation-label';
+          deprecationLabel.innerText = 'Deprecated';
+          deprecation.appendChild(deprecationLabel);
+
+          deprecationReason = document.createElement('div');
+          deprecationReason.className =
+            'CodeMirror-hint-information-deprecation-reason';
+          deprecation.appendChild(deprecationReason);
+
+          /**
+           * This is a bit hacky: By default, codemirror renders all hints
```

Comment:
That was probably the trickiest part: Getting the two-column layout working for the typeahead hints. I hope this comment explains how I made this work with the restrictions imposed by codemirror.

Also, if you see a better way, happy to change the implementation!